src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.pubsublite.flink;
+
+import com.google.api.core.ApiFutures;
+import com.google.cloud.pubsublite.AdminClient;
+import com.google.cloud.pubsublite.Offset;
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import com.google.cloud.pubsublite.TopicPath;
+import com.google.cloud.pubsublite.flink.internal.reader.PartitionFinishedCondition;
+import com.google.cloud.pubsublite.flink.internal.reader.PartitionFinishedCondition.Factory;
+import com.google.cloud.pubsublite.flink.internal.reader.PartitionFinishedCondition.Result;
+import com.google.cloud.pubsublite.internal.ExtractStatus;
+import com.google.cloud.pubsublite.internal.TopicStatsClient;
+import com.google.cloud.pubsublite.proto.Cursor;
+import com.google.common.collect.ImmutableMap;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+public abstract class StopCondition implements Serializable {
+  // Called to convert conditions with placeholder values like "HEAD" to an offset based condition.
+  abstract StopCondition canonicalize(
+      SubscriptionPath path,
+      Supplier<AdminClient> adminClient,
+      Supplier<TopicStatsClient> topicStatsClient);
+
+  abstract PartitionFinishedCondition.Factory toFinishCondition();
+
+  private static class ContinueIndefinitely extends StopCondition {
+    @Override
+    public StopCondition canonicalize(
+        SubscriptionPath path,
+        Supplier<AdminClient> adminClient,
+        Supplier<TopicStatsClient> topicStatsClient) {
+      return this;
+    }
+
+    @Override
+    public Factory toFinishCondition() {
+      return (subscription, partition) -> offset -> Result.CONTINUE;
+    }
+  }
+
+  private static class ReadToOffsets extends StopCondition {
+    private final Map<Partition, Offset> offsets;
+
+    private ReadToOffsets(Map<Partition, Offset> offsets) {
+      this.offsets = offsets;
+    }
+
+    @Override
+    public StopCondition canonicalize(
```

Comment:
I think you could provide a default implementation to return this, and only override ReadToHead.

Reply:
Done